changefeedccl: add SASL authentication for Kafka

rolandcrosby opened the pull request:
Adds support for SASL/PLAIN authentication via Kafka URL parameters.
This was already supported by the Sarama library we use for Kafka, so
this change just plumbs the parameters through to Sarama.

I've tested this with and without TLS against a Confluent cluster with
SASL; I have a reproducible script to run these tests in a Roachprod
cluster, but don't know how I'd go about adding such a test to our
automation.

As discussed with Dan, this likely will need to be backported into 19.1.1
at this point, which should be fine as far as customer commitments go. 

Release note (enterprise change): `CHANGEFEED`s now support SASL/PLAIN
authentication when connecting to a Kafka sink.

Commit:
changefeedccl: add SASL authentication for Kafka

Adds support for SASL/PLAIN authentication via Kafka URL parameters.
This was already supported by the Sarama library we use for Kafka, so
this change just plumbs the parameters through Sarama.

Newly added parameters are:
- `sasl_enabled` - boolean, whether to use SASL auth
- `sasl_handshake` - boolean, on by default, used in certain SASL
    proxy configurations
- `sasl_user` - SASL/PLAIN username
- `sasl_password` - SASL/PLAIN password

I've tested this with and without TLS against a Confluent cluster with
SASL; I have a reproducible script to run these tests in a Roachprod
cluster, but don't know how I'd go about adding such a test to our
automation.

Release note (enterprise change): `CHANGEFEED`s now support SASL/PLAIN
authentication when connecting to a Kafka sink.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35800)
<!-- Reviewable:end -->

danhhz reviewed (commented):
Document the params in the commit message please.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35800#-:-La1qewn7UgxLHn4MiF2:b2oxiwx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rolandcrosby)

___
*[pkg/ccl/changefeedccl/sink.go, line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35800#-La1qieQ3NYSDimlAKEf:-La1qieQ3NYSDimlAKEg:bo13xtn) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f9f2b915510f068cd4518259364725b794941e0/pkg/ccl/changefeedccl/sink.go#L138)):*
> ```Go
> 		if cfg.saslEnabled {
> 			if cfg.saslUser == `` {
> 				return nil, errors.Errorf(`%s must be provided when SASL is enabled`, sinkParamSASLUser)
> ```

Add some assertions to TestChangefeedErrors in changefeed_test.go for the various interesting combinations of these, please. Feel free to pattern match https://github.com/cockroachdb/cockroach/blob/b4b081d307a3bc9fc1e29b4448cfe071b8452741/pkg/ccl/changefeedccl/changefeed_test.go#L1391-L1403
___
*[pkg/ccl/changefeedccl/sink.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35800#-La1qU1262f8pXSasehL:-La1qU1262f8pXSasehM:b-2ndoj1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f9f2b915510f068cd4518259364725b794941e0/pkg/ccl/changefeedccl/sink.go#L144)):*
> ```Go
> 			}
> 		} else {
> 			if cfg.saslUser != `` {
> ```

should this also check handshake? guessing not since it defaults to true


<!-- Sent from Reviewable.io -->

rolandcrosby reviewed (commented):
TFTR! Added params to the commit message and addressed feedback inline.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35800#-:-La1xWQB38_ONCvrIegD:b-iu9x3w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz)

___
*[pkg/ccl/changefeedccl/sink.go, line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35800#-La1qieQ3NYSDimlAKEf:-La1vVaTEv4Wv7Xo0kEO:by1i31v) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f9f2b915510f068cd4518259364725b794941e0/pkg/ccl/changefeedccl/sink.go#L138)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Add some assertions to TestChangefeedErrors in changefeed_test.go for the various interesting combinations of these, please. Feel free to pattern match https://github.com/cockroachdb/cockroach/blob/b4b081d307a3bc9fc1e29b4448cfe071b8452741/pkg/ccl/changefeedccl/changefeed_test.go#L1391-L1403
</blockquote></details>

Done. Let me know if I missed anything that looks interesting.
___
*[pkg/ccl/changefeedccl/sink.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35800#-La1qU1262f8pXSasehL:-La1v_UVDRb-MYjpAtKm:b-gv48lb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f9f2b915510f068cd4518259364725b794941e0/pkg/ccl/changefeedccl/sink.go#L144)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

should this also check handshake? guessing not since it defaults to true
</blockquote></details>

Added a new check for completeness at the time the parameter is parsed.


<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Looks like you have some compile errors in the test. You can run it locally with `make build PKG=./pkg/ccl/changefeedccl TESTS=ChangefeedErrors`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35800#-:-La2WxF6BOO-u4z3-BRQ:b4cosf7)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz)



<!-- Sent from Reviewable.io -->

rolandcrosby commented:
Fixed the issues in the test cases I added, but now (if I'm reading this correctly) TeamCity is reporting logictest failures that I can't repro locally.

bors r=danhhz
